# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [2]:
import pandas as pd
import numpy as np

Carregamento do arquivo

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
dataset = pd.read_csv('/content/drive/MyDrive/Colab_Notebooks/EBAC/Cientista de Dados/Módulo 7 - Árvores de Decisão/demo01.csv')

As variáveis e seus tipos



In [5]:
dataset.dtypes

,0
sexo,object
posse_de_veiculo,object
posse_de_imovel,object
qtd_filhos,int64
tipo_renda,object
educacao,object
estado_civil,object
tipo_residencia,object
idade,float64
tempo_emprego,float64


In [6]:
dataset.head()

,sexo,posse_de_veiculo,posse_de_imovel,qtd_filhos,tipo_renda,educacao,estado_civil,tipo_residencia,idade,tempo_emprego,possui_celular,possui_fone_comercial,possui_fone,possui_email,qt_pessoas_residencia,mau
0,M,Y,Y,0,Working,Secondary / secondary special,Married,House / apartment,58.832877,3.106849,1,0,0,0,2.0,False
1,F,N,Y,0,Commercial associate,Secondary / secondary special,Single / not married,House / apartment,52.356164,8.358904,1,0,1,1,1.0,False
2,F,N,Y,0,Commercial associate,Secondary / secondary special,Single / not married,House / apartment,52.356164,8.358904,1,0,1,1,1.0,False
3,M,Y,Y,0,Working,Higher education,Married,House / apartment,46.224658,2.106849,1,1,1,1,2.0,False
4,F,Y,N,0,Working,Incomplete higher,Married,House / apartment,29.230137,3.021918,1,0,0,0,2.0,False


Quantidade de valores nulos

In [7]:
dataset.isna().sum()

,0
sexo,0
posse_de_veiculo,0
posse_de_imovel,0
qtd_filhos,0
tipo_renda,0
educacao,0
estado_civil,0
tipo_residencia,0
idade,0
tempo_emprego,0


In [8]:
for col in dataset.columns:
  print('=' * 40)
  print(dataset[col].value_counts())

sexo
F    11201
M     5449
Name: count, dtype: int64
posse_de_veiculo
N    10178
Y     6472
Name: count, dtype: int64
posse_de_imovel
Y    11176
N     5474
Name: count, dtype: int64
qtd_filhos
0     11486
1      3393
2      1552
3       189
4        24
5         2
14        2
7         2
Name: count, dtype: int64
tipo_renda
Working                 8565
Commercial associate    3826
Pensioner               2800
State servant           1451
Student                    8
Name: count, dtype: int64
educacao
Secondary / secondary special    11245
Higher education                  4551
Incomplete higher                  649
Lower secondary                    188
Academic degree                     17
Name: count, dtype: int64
estado_civil
Married                 11680
Single / not married     2035
Civil marriage           1283
Separated                 945
Widow                     707
Name: count, dtype: int64
tipo_residencia
House / apartment      14974
With parents             738
Municipal 

In [9]:
dataset.mau.value_counts()

,count
mau,
False,16260
True,390


#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.
2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável.
    Dica:
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis,
        3. conte o número de categorias dessa variável
        4. acumule essa informação de 3. na lista que você criou em 1.
        5. No final, essa lista pode ser atribuída à nossa variável.

In [27]:
metadata = {coluna: dataset[coluna].dtype for coluna in dataset.columns}

In [28]:
metadata

{'sexo': dtype('O'),
 'posse_de_veiculo': dtype('O'),
 'posse_de_imovel': dtype('O'),
 'qtd_filhos': dtype('int64'),
 'tipo_renda': dtype('O'),
 'educacao': dtype('O'),
 'estado_civil': dtype('O'),
 'tipo_residencia': dtype('O'),
 'idade': dtype('float64'),
 'tempo_emprego': dtype('float64'),
 'possui_celular': dtype('int64'),
 'possui_fone_comercial': dtype('int64'),
 'possui_fone': dtype('int64'),
 'possui_email': dtype('int64'),
 'qt_pessoas_residencia': dtype('float64'),
 'mau': dtype('bool')}

In [29]:
metadata = pd.DataFrame({'nome_variavel': metadata.keys(), 'tipo_dado': metadata.values()})
categorias = []
for coluna in dataset.columns:
    if dataset[coluna].dtype == 'object':
        categorias.append(dataset[coluna].nunique())
    else:
        categorias.append(0)
metadata['qtd_categorias'] = categorias
display(metadata)

,nome_variavel,tipo_dado,qtd_categorias
0,sexo,object,2
1,posse_de_veiculo,object,2
2,posse_de_imovel,object,2
3,qtd_filhos,int64,0
4,tipo_renda,object,5
5,educacao,object,5
6,estado_civil,object,5
7,tipo_residencia,object,6
8,idade,float64,0
9,tempo_emprego,float64,0


#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [30]:
variaveis_categoricas = [categorica for categorica in dataset.columns if dataset[categorica].dtype == 'object']
dataset_encoded = pd.get_dummies(dataset, columns=variaveis_categoricas, drop_first=True)
display(dataset_encoded)

,qtd_filhos,idade,tempo_emprego,possui_celular,possui_fone_comercial,possui_fone,possui_email,qt_pessoas_residencia,mau,sexo_M,...,educacao_Secondary / secondary special,estado_civil_Married,estado_civil_Separated,estado_civil_Single / not married,estado_civil_Widow,tipo_residencia_House / apartment,tipo_residencia_Municipal apartment,tipo_residencia_Office apartment,tipo_residencia_Rented apartment,tipo_residencia_With parents
0,0,58.832877,3.106849,1,0,0,0,2.0,False,True,...,True,True,False,False,False,True,False,False,False,False
1,0,52.356164,8.358904,1,0,1,1,1.0,False,False,...,True,False,False,True,False,True,False,False,False,False
2,0,52.356164,8.358904,1,0,1,1,1.0,False,False,...,True,False,False,True,False,True,False,False,False,False
3,0,46.224658,2.106849,1,1,1,1,2.0,False,True,...,False,True,False,False,False,True,False,False,False,False
4,0,29.230137,3.021918,1,0,0,0,2.0,False,False,...,False,True,False,False,False,True,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16645,0,54.109589,9.884932,1,0,0,0,2.0,True,False,...,True,False,False,False,False,True,False,False,False,False
16646,0,43.389041,7.380822,1,1,1,0,2.0,True,False,...,True,True,False,False,False,True,False,False,False,False
16647,0,30.005479,9.800000,1,1,0,0,2.0,True,True,...,True,True,False,False,False,True,False,False,False,False
16648,0,30.005479,9.800000,1,1,0,0,2.0,True,True,...,True,True,False,False,False,True,False,False,False,False


In [32]:
for coluna in dataset_encoded.columns:
  if dataset_encoded[coluna].dtype == 'bool':
    dataset_encoded[coluna] = dataset_encoded[coluna].astype(int)

In [36]:
dataset_encoded.qt_pessoas_residencia = dataset_encoded.qt_pessoas_residencia.astype(int)

In [37]:
dataset_encoded

,qtd_filhos,idade,tempo_emprego,possui_celular,possui_fone_comercial,possui_fone,possui_email,qt_pessoas_residencia,mau,sexo_M,...,educacao_Secondary / secondary special,estado_civil_Married,estado_civil_Separated,estado_civil_Single / not married,estado_civil_Widow,tipo_residencia_House / apartment,tipo_residencia_Municipal apartment,tipo_residencia_Office apartment,tipo_residencia_Rented apartment,tipo_residencia_With parents
0,0,58.832877,3.106849,1,0,0,0,2,0,1,...,1,1,0,0,0,1,0,0,0,0
1,0,52.356164,8.358904,1,0,1,1,1,0,0,...,1,0,0,1,0,1,0,0,0,0
2,0,52.356164,8.358904,1,0,1,1,1,0,0,...,1,0,0,1,0,1,0,0,0,0
3,0,46.224658,2.106849,1,1,1,1,2,0,1,...,0,1,0,0,0,1,0,0,0,0
4,0,29.230137,3.021918,1,0,0,0,2,0,0,...,0,1,0,0,0,1,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16645,0,54.109589,9.884932,1,0,0,0,2,1,0,...,1,0,0,0,0,1,0,0,0,0
16646,0,43.389041,7.380822,1,1,1,0,2,1,0,...,1,1,0,0,0,1,0,0,0,0
16647,0,30.005479,9.800000,1,1,0,0,2,1,1,...,1,1,0,0,0,1,0,0,0,0
16648,0,30.005479,9.800000,1,1,0,0,2,1,1,...,1,1,0,0,0,1,0,0,0,0


#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

In [38]:
# Cross-tabulation for 'possui_email' and 'mau'
crosstab_email = pd.crosstab(dataset_encoded['possui_email'], dataset_encoded['mau'], normalize='index')
print("Tabela cruzada: possui_email vs. mau")
display(crosstab_email)

# Probability of 'mau = 1' for 'possui_email'
prob_mau_email = crosstab_email[1].sort_values(ascending=False)
print("\nProbabilidade de mau = 1 por possui_email:")
display(prob_mau_email)

# Cross-tabulation for 'posse_de_veiculo_Y' (as 'posse_de_veiculo' is dummified) and 'mau'
crosstab_veiculo = pd.crosstab(dataset_encoded['posse_de_veiculo_Y'], dataset_encoded['mau'], normalize='index')
print("\nTabela cruzada: posse_de_veiculo_Y vs. mau")
display(crosstab_veiculo)

# Probability of 'mau = 1' for 'posse_de_veiculo_Y'
prob_mau_veiculo = crosstab_veiculo[1].sort_values(ascending=False)
print("\nProbabilidade de mau = 1 por posse_de_veiculo_Y:")
display(prob_mau_veiculo)

Tabela cruzada: possui_email vs. mau


mau,0,1
possui_email,,
0,0.976269,0.023731
1,0.979730,0.020270



Probabilidade de mau = 1 por possui_email:


,1
possui_email,
0,0.023731
1,0.020270



Tabela cruzada: posse_de_veiculo_Y vs. mau


mau,0,1
posse_de_veiculo_Y,,
0,0.975732,0.024268
1,0.977905,0.022095



Probabilidade de mau = 1 por posse_de_veiculo_Y:


,1
posse_de_veiculo_Y,
0,0.024268
1,0.022095


Ao analisar as tabelas de probabilidade de `mau = 1`:

*   **Possui Email:**
    *   Para quem *não possui email* (`possui_email = 0`), a probabilidade de `mau = 1` é de aproximadamente **2.38%**.
    *   Para quem *possui email* (`possui_email = 1`), a probabilidade de `mau = 1` é de aproximadamente **2.03%**.

*   **Possui Veículo:**
    *   Para quem *não possui veículo* (`posse_de_veiculo_Y = 0`, ou seja, 'N'), a probabilidade de `mau = 1` é de aproximadamente **2.43%**.
    *   Para quem *possui veículo* (`posse_de_veiculo_Y = 1`, ou seja, 'Y'), a probabilidade de `mau = 1` é de aproximadamente **2.21%**.

A variável `posse_de_veiculo` (ou `posse_de_veiculo_Y` após a dummyficação) parece ser um pouco mais poderosa para prever a probabilidade de `mau = 1` do que `possui_email`. Há uma diferença percentual maior na probabilidade de `mau = 1` entre os grupos que possuem e não possuem veículo (aproximadamente 0.67 pontos percentuais de diferença) em comparação com os grupos que possuem e não possuem email (aproximadamente 0.08 pontos percentuais de diferença). Isso sugere que a `posse_de_veiculo` tem uma separação mais distinta nos resultados de `mau = 1`.

#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [39]:
dataset_encoded.to_csv('/content/drive/MyDrive/Colab_Notebooks/EBAC/Cientista de Dados/Módulo 7 - Árvores de Decisão/demo01_encoded.csv', index=False)